core/shared/src/main/scala/fs2/retry.scala

```diff
@@ -0,0 +1,57 @@
+package fs2
+
+import cats.effect.Async
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+object retry {
+  /**
+   * Retries `fa` on failure, returning a single element
+   * stream with the result of `fa` as soon as it succeeds.
+   *
+   * @param delay Duration of delay before the first retry
+   *
+   * @param nextDelay Applied to the previous delay to compute the
+   *                  next, e.g. to implement exponential backoff
+   *
+   * @param maxRetries Number of attempts before failing with the
+   *                   latest error, if `fa` never succeeds
+   *
+   * @param retriable Function to determine whether a failure is
+   *                  retriable or not, defaults to retry every
+   *                  `NonFatal`. A failed stream is immediately
+   *                  returned when a non-retriable failure is
+   *                  encountered
+   */
+  def retry[F[_], A](fa: F[A],
+                     delay: FiniteDuration,
```

Comment:
Thanks for the feedback :) 
A few comments:
- `F` vs `Stream`: the counterargument is that (I think) you would need to guarantee that a stream only returns one result, which didn't seem super clean.
- `retryWhen`: I actually started with a simpler recursive method using `flatMap` (as opposed to combinators over an infinite stream), along the lines of:
```scala
def retryWithDelay[F[_], A](
    fa: F[A],
    delay: FiniteDuration,
    nextDelay: FiniteDuration => FiniteDuration,
    maxRetries: Int
  )(implicit sch: Scheduler, F: Async[F]): Stream[F, A] =
    Stream.attemptEval(fa).flatMap {
      case Left(ex) if maxRetries <= 0 => Stream.fail(ex)
      case Left(ex)  => fs2.time.sleep_(delay) ++ retryWithDelay(fa, nextDelay(delay), nextDelay, maxRetries.map(_ - 1))
      case Right(a) => Stream.emit(a)
    }
```
 and you seem to suggest a similar approach (correct me if I'm wrong). The reason I moved away from it is that standard `Stream` combinators didn't seem to work (e.g. `take`), which I think is counterintuitive.
- Parameters: They roughly mirror the existing `retry` in Scala.Task (which several people have asked for a replacement of). Definitely open to a discussion over those though :)